### PR TITLE
Solo Env / PyBullet Encapsulation

### LINKED ISSUE #18: Forcing One PyBullet Client per Environment
I guess that we didn't know how to PyBullet worked when we started. When trying to train the models in a env and then creating a new gui-based environment for rendering, we found that we had run into a lot of issues.

The only way to circumvent them was to restart the kernel, which also happens to be the only way to terminate the pybullet physics server. I'm gonna conjecture that when we create a second environment, it causes conflicts because there already is a simulation running on the PyBullet server. I believe if we forced each environment to maintain its own physics client, then we should be a-ok.

It's either that or find a way to train without having to render and be able to start rendering mid-pybullet session.

### PR DESCRIPTION
**Pull request by agupta231**
Closes #18.

Since PyBullet's GUI functionality is dependent upon the initialization, the best way to train + demo a net would be the following:
1. Create a non-gui env
2. Train the model on that for speed
3. Create a gui env
4. Perform demonstrations on that environment

Previously, all of the gym environments had shared the global PyBullet pointer, so all of the calculations ended up conflicting. Thus the only way to render the environment would be to save / load the model between a training script and a rendering script. This PR makes it so that everything can be done in one script.

This functionality has been tested + working empirically in a Jupyter notebook.

@mahajanrevant, watch out, this is also going to affect how your `Termination` objects work as well in #27 and #19.

### CODE REVIEW

File: gym_solo/core/obs.py [hunk 1]
```diff
@@ -13,6 +14,10 @@
 
 
 class Observation(ABC):
+  """An observation for a body in the pybullet simulation.
+  """
+  _client: bullet_client.BulletClient = None
```
Comment (mahajanrevant) [inline gym_solo/core/obs.py:19]: A little context on why is this change needed?

Comment (mahajanrevant) [reply gym_solo/core/obs.py:19]: I understood from how you are using it but it would be nice to have a one-liner comment here

File: gym_solo/core/obs.py [hunk 2]
```diff
@@ -222,7 +258,10 @@ def __init__(self, body_id: int, degrees: bool = False):
     """
     self.robot = body_id
     self._degrees = degrees
-    self._num_joints = p.getNumJoints(self.robot)
+
+  @property
```
Comment (mahajanrevant) [inline gym_solo/core/obs.py:262]: This is just for my own curiosity. Is it a good practice to declare most/all methods as property?

File: gym_solo/core/test_obs_observations.py [hunk 3]
```diff
@@ -174,15 +199,15 @@ def test_labels(self, mock_num_joints, mock_joint_info):
...
-    self.assertEqual(o.labels(), ground_truth)
+    o = self.build_obs(dummy_robot_id)
+    self.assertEqual(o.labels, ground_truth)
     
   @parameterized.expand([
     ("default", False),
     ("degrees", True)
   ])
-  @mock.patch('pybullet.getJointState', autospec=True)
-  @mock.patch('pybullet.getNumJoints', autospec=True)
```
Comment (mahajanrevant) [inline gym_solo/core/test_obs_observations.py:185]: What does removing autospec do?

File: gym_solo/core/test_obs_observations.py [hunk 4]
```diff
@@ -1,20 +1,44 @@
...
+    pass
+
+  def setUp(self):
+    self.client = bullet_client.BulletClient(connection_mode=p.DIRECT)
+
+  def tearDown(self):
+    self.client.disconnect()
+
+  def build_obs(self, *args, **kwargs):
+    o = self.obs_cls(*args, **kwargs)
```
Comment (mahajanrevant) [inline gym_solo/core/test_obs_observations.py:28]: obs_cls does not allow for any arguments correct? Are you expecting subclasses will take care of it?

File: gym_solo/envs/test_solo8v2vanilla.py [hunk 5]
```diff
@@ -144,5 +146,23 @@ def test_observation_space(self):
...
+  def test_disjoint_environments(self):
+    env1 = solo_env.Solo8VanillaEnv(config=solo_env.Solo8VanillaConfig())
+    env1.obs_factory.register_observation(solo_obs.TorsoIMU(env1.robot))
+    env1.obs_factory.register_observation(solo_obs.MotorEncoder(env1.robot))
+    env1.reward_factory.register_reward(1, SimpleReward())
+
+    home_position = env1.reset()
+    
+    for i in range(1000):
+      env1.step(env1.action_space.sample())
```
Comment (mahajanrevant) [inline gym_solo/envs/test_solo8v2vanilla.py:158]: Why do we need this for loop? There is nothing being tested from env1 except the home position

File: gym_solo/core/obs.py [hunk 6]
```diff
@@ -58,11 +63,40 @@ def compute(self) -> solo_types.obs:
...
+
+    Returns:
+      bullet_client.BulletClient: The active client for the observation.
+    """
+    if not self._client:
+      raise ValueError('PyBullet client needs to be set')
+    return self._client
+
+  @client.setter
+  def client(self, client: bullet_client.BulletClient):
```
[the remark this answers is not in the document]
Comment (agupta231) [reply gym_solo/core/obs.py:81]: it's not [pythonic](https://stackoverflow.com/questions/2627002/whats-the-pythonic-way-to-use-getters-and-setters)

File: gym_solo/core/obs.py (hunk 1 above)
Comment (agupta231) [reply gym_solo/core/obs.py:19]: done

File: gym_solo/core/obs.py (hunk 2 above)
Comment (agupta231) [reply gym_solo/core/obs.py:262]: so python is more about the "semantic" meaning behind the attribute/method. Methods are meant to _do_ things while attribute are meant to _be_ things. That's usually why it's a code style to force method names to be verbs. 

In the case of `_num_joints`, it _is_ something rather than _doing_ something; therefore it should be an attribute. However, we know that it depends upon the client to compute the value, which isn't available at instantiation. As per the API requirements, we know that the `client` will be set at registration time, and that `_num_joints` will only be computed in either the `observation_space`, `labels`, or, `compute` functionalities. Therefore, we know that `_num_joints` will only be accessed after the `Observation` has been registered.

Thus, since we need access to the client, we can "fake" this attribute into a method using `@property` so that we can evaluate it lazily. However, its still accessed as `Observation._num_joints` so it's semantically still an attribute. For something like this, I'd use [functool's `@cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property) for efficiency's sake, but that's unfortunately py3.8+. Since this is just a simple lookup on PyBullet's end, I'm not losing any sleep about the overhead.

File: gym_solo/core/test_obs_observations.py (hunk 3 above)
Comment (agupta231) [reply gym_solo/core/test_obs_observations.py:185]: `autospec` requires that the mock has the _exact_ same method signature as the method it is patching. However, since we are using a pybullet client wrapper now, we have extra variables being passed through. Removing autospec lets mock handle everything and makes tests a bit easier.

File: gym_solo/core/test_obs_observations.py (hunk 4 above)
Comment (agupta231) [reply gym_solo/core/test_obs_observations.py:28]: since it's an abstract property, it's not actually a method. Rather, it's a `Callable` class. The arguments are the input arguments to the `Observation`. Like in the subclass:

https://github.com/WPI-MMR/gym_solo/blob/e7c17df8bf0a2011d9e9dff7efb5734fde142347/gym_solo/core/test_obs_observations.py#L33-L34

The arguments are the input arguments to `TorsoIMU`

File: gym_solo/envs/test_solo8v2vanilla.py (hunk 5 above)
Comment (agupta231) [reply gym_solo/envs/test_solo8v2vanilla.py:158]: to ensure that `env1` isn't in the home position. then, if `env2`'s observation is the home position, we know that `env1` and `env2` are fully disjoint.

File: gym_solo/core/obs.py (hunk 2 above)
Comment (mahajanrevant) [reply gym_solo/core/obs.py:266]: That's an awesome explanation! Thank you dad!

**Review by mahajanrevant — APPROVED** [review]
Looks good
